File: Medium/number_of_islands/number_of_islands.py
from typing import List

def numIslands(grid: List[List[str]]) -> int:
    if not grid:
        return 0
    rows = len(grid)
    cols = len(grid[0])
    count = 0

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == "1":
                count += 1
                stack = []
                stack.append([r, c])
                while len(stack) > 0:
                    cell = stack.pop()
                    current_row = cell[0]
                    current_col = cell[1]

                    if grid[current_row][current_col] == "0":
                        continue
                    grid[current_row][current_col] = "0"

                    new_r = current_row - 1
                    new_c = current_col
                    if new_r >= 0:                
                        if grid[new_r][new_c] == "1":
                            stack.append([new_r, new_c])

                    new_r = current_row + 1
                    new_c = current_col
                    if new_r < rows:               
                        if grid[new_r][new_c] == "1":
                            stack.append([new_r, new_c])

                    new_r = current_row
                    new_c = current_col - 1
                    if new_c >= 0:                 
                        if grid[new_r][new_c] == "1":
                            stack.append([new_r, new_c])

                    new_r = current_row
                    new_c = current_col + 1
                    if new_c < cols:               
                        if grid[new_r][new_c] == "1":
                            stack.append([new_r, new_c])
    return count
    
if __name__ == "__main__":
    print("Testing numIslands function:")
    grid1 = [
      ["1","1","1","1","0"],
      ["1","1","0","1","0"],
      ["1","1","0","0","0"],
      ["0","0","0","0","0"]
    ]
    print("For the input grid1, the number of islands is:", numIslands(grid1)) #Output: 1

    grid2 = [
      ["1","1","0","0","0"],
      ["1","1","0","0","0"],
      ["0","0","1","0","0"],
      ["0","0","0","1","1"]
    ]
    print("For the input grid2, the number of islands is:", numIslands(grid2)) #Output: 3

File: Medium/number_of_islands/number_of_islands.cpp
#include <vector>
#include <stack>
#include <iostream>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        if(rows == 0){
            return 0;
        }
        int cols = grid[0].size();
        int islandCount = 0;

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == '1') {
                    islandCount++;

                    //Using vector<int> to store coordinates manually
                    stack<vector<int>> s;
                    s.push({i, j});

                    while(!s.empty()) {
                        vector<int> cell = s.top();
                        s.pop();
                        int r = cell[0];
                        int c = cell[1];

                        if(grid[r][c] == '0'){
                            continue;
                        }
                        grid[r][c] = '0'; // mark as visited

                        // Push neighbors if valid land
                        if(r > 0 && grid[r-1][c] == '1'){
                            s.push({r-1, c});
                        }
                        if(r < rows-1 && grid[r+1][c] == '1'){
                            s.push({r+1, c});
                        }
                        if(c > 0 && grid[r][c-1] == '1'){
                            s.push({r, c-1});
                        }
                        if(c < cols-1 && grid[r][c+1] == '1'){
                            s.push({r, c+1});
                        }
                    }
                }
            }
        }
        return islandCount;
    }
};

//Example usage
int main() {
    Solution solution1;
    vector<vector<char>> grid1 = {
        {'1','1','1','1','0'},
        {'1','1','0','1','0'},
        {'1','1','0','0','0'},
        {'0','0','0','0','0'}
    };
    cout << "Number of islands: " << solution1.numIslands(grid1) << endl;

    Solution solution2;
    vector<vector<char>> grid2 = {
        {'1','1','0','0','0'},
        {'1','1','0','0','0'},
        {'0','0','1','0','0'},
        {'0','0','0','1','1'}
    };
    cout << "Number of islands: " << solution2.numIslands(grid2) << endl;
    return 0;
}